### ISSUE TITLE
[Feature Request] Fine grained precision behavior control of torch.compile

### ISSUE THREAD

**Issue by jerrychenhf** [pytorch/pytorch#196957]
### 🚀 The feature, motivation and pitch

When we are debugging an accuracy issue from FLUX pipeline, we found that there are some layers of the model are prone to numeric drift causing by torch.compile when operations are fused with a higher precision. 

For example, we found that time_text_embed layer of FluxTransformer2DModel is such a layer. Below is the skeleton of the forward function of FluxTransformer2DModel I used for debugging. If we compile the full model running on BFloat16, we found that results of time_text_embed gets very large numeric drift comparing with eager. 

```
def forward(
        self,
        hidden_states: torch.Tensor,
        encoder_hidden_states: torch.Tensor,
        pooled_projections: torch.Tensor,
        timestep: torch.Tensor,
        img_ids: torch.Tensor,
        txt_ids: torch.Tensor,
        guidance: torch.Tensor | None = None,
        joint_attention_kwargs: dict[str, Any] | None = None,
    ) -> tuple[
        torch.Tensor,
        torch.Tensor,
        torch.Tensor,
        torch.Tensor,
        torch.Tensor,
        torch.Tensor,
        torch.Tensor,
    ]:
        t = self.transformer

        hidden_states = t.x_embedder(hidden_states)
        x_embed_out = hidden_states

        timestep = timestep.to(hidden_states.dtype) * 1000
        if guidance is not None:
            guidance = guidance.to(hidden_states.dtype) * 1000

        temb = (
                t.time_text_embed(timestep, pooled_projections)
                if guidance is None
                else t.time_text_embed(timestep, guidance, pooled_projections)
            )
        encoder_hidden_states = t.context_embedder(encoder_hidden_states)

        if txt_ids.ndim == 3:
            txt_ids = txt_ids[0]
        if img_ids.ndim == 3:
            img_ids = img_ids[0]

        ids = torch.cat((txt_ids, img_ids), dim=0)
        image_rotary_emb = t.pos_embed(ids)

        local_kwargs = dict(joint_attention_kwargs or {})
        if "ip_adapter_image_embeds" in local_kwargs:
            ip_adapter_image_embeds = local_kwargs.pop("ip_adapter_image_embeds")
            ip_hidden_states = t.encoder_hid_proj(ip_adapter_image_embeds)
            local_kwargs.update({"ip_hidden_states": ip_hidden_states})

        # Skip transformer_blocks and single_transformer_blocks on purpose.
        hidden_states = t.norm_out(hidden_states, temb)
        norm_out_out = hidden_states
        output = t.proj_out(hidden_states)

        return (
            x_embed_out,
            temb,
            norm_out_out,
            output,
            encoder_hidden_states,
            image_rotary_emb[0],
            image_rotary_emb[1],
        )

```
If we add a graph break before temb= time_text_embed layer, temb numeric drift of torch compile getting back to normal.

We created script to reproduce the behavior:

[repro_time_text_embed_compile_diff.py](https://github.com/user-attachments/files/32182887/repro_time_text_embed_compile_diff.py)

Running the script:
`python3 repro_time_text_embed_compile_diff.py --device cpu --dtypes bfloat16 --batch-size 1 --image-seq-len 8 --in-channels 64 --embedding-dim 64 --pooled-projection-dim 128 --timestep-scale 1000`

It show very big numeric drift on stage_time_text_embed (max_abs=2.822266e-01 mean_abs=7.722092e-02 rmse=9.964155e-02 rel_l2=3.929916e-01): 

```
[RUN] dtype=bfloat16
  [SCENARIO] name=single_run graph_break=False
    stage_x_embedder: close=True exact=True max_abs=0.000000e+00 mean_abs=0.000000e+00 rmse=0.000000e+00 rel_l2=0.000000e+00
    stage_time_text_embed: close=False exact=False max_abs=2.822266e-01 mean_abs=7.722092e-02 rmse=9.964155e-02 rel_l2=3.929916e-01
    stage_norm_out: close=False exact=False max_abs=2.343750e-01 mean_abs=2.697963e-02 rmse=3.687467e-02 rel_l2=3.687517e-02
    stage_proj_out: close=False exact=False max_abs=6.738281e-02 mean_abs=1.565552e-02 rmse=2.008618e-02 rel_l2=3.435689e-02
  [RESULT] PASS=False atol=0.01 rtol=0.01
```

Either graph break before time_text_embed layer or set torch._inductor.config.emulate_precision_casts=True fix the problem. 

`python3 repro_time_text_embed_compile_diff.py --device cpu --dtypes bfloat16 --batch-size 1 --image-seq-len 8 --in-channels 64 --embedding-dim 64 --pooled-projection-dim 128 --timestep-scale 1000 --graph-break-around-time-text-embed`

or 

`python3 repro_time_text_embed_compile_diff.py --device cpu --dtypes bfloat16 --batch-size 1 --image-seq-len 8 --in-channels 64 --embedding-dim 64 --pooled-projection-dim 128 --timestep-scale 1000 --emulate-precision-casts`

The huge numeric drift is caused by operation fusions running on float32:

```
                    auto tmp0 = in_ptr0[static_cast<int64_t>(0L)];
                    auto tmp1 = c10::convert<float>(tmp0);
                    auto tmp2 = static_cast<float>(1000.0);
                    auto tmp3 = float(tmp1 * tmp2);
                    auto tmp4 = c10::convert<float>(tmp3);
                    auto tmp5 = x0;
                    auto tmp6 = c10::convert<float>(tmp5);
                    auto tmp7 = at::vec::Vectorized<float>::arange(tmp6, 1);
                    auto tmp8 = static_cast<float>(-9.210340371976184);
                    auto tmp9 = at::vec::Vectorized<float>(tmp8);
                    auto tmp10 = tmp7 * tmp9;
                    auto tmp11 = static_cast<float>(128.0);
                    auto tmp12 = at::vec::Vectorized<float>(tmp11);
                    auto tmp13 = tmp10 / tmp12;
                    auto tmp14 = tmp13.exp();
                    auto tmp15 = at::vec::Vectorized<float>(tmp4);
                    auto tmp16 = tmp15 * tmp14;
                    auto tmp17 = tmp16.sin();
                    auto tmp18 = tmp16.cos();
                    tmp17.store(out_ptr1 + static_cast<int64_t>(x0));
                    tmp18.store(out_ptr2 + static_cast<int64_t>(x0));
```
While at eager, or with a graph break or emulate-precision-casts, the result of tmp3 is cast to BFloat16 and back to float32 to do sin, cos operations which may non-linearly increase the drift.

```
                    auto tmp0 = in_ptr0[static_cast<int64_t>(0L)];
                    auto tmp1 = c10::convert<at::BFloat16>(tmp0);
                    auto tmp2 = c10::convert<float>(tmp1);
                    auto tmp3 = static_cast<float>(1000.0);
                    auto tmp4 = float(tmp2 * tmp3);
                    auto tmp5 = c10::convert<at::BFloat16>(tmp4);
                    auto tmp6 = c10::convert<float>(tmp5);
                    auto tmp7 = x0;
                    auto tmp8 = c10::convert<float>(tmp7);
                    auto tmp9 = at::vec::Vectorized<float>::arange(tmp8, 1);
                    auto tmp10 = static_cast<float>(-9.210340371976184);
                    auto tmp11 = at::vec::Vectorized<float>(tmp10);
                    auto tmp12 = tmp9 * tmp11;
                    auto tmp13 = static_cast<float>(128.0);
                    auto tmp14 = at::vec::Vectorized<float>(tmp13);
                    auto tmp15 = tmp12 / tmp14;
                    auto tmp16 = tmp15.exp();
                    auto tmp17 = at::vec::Vectorized<float>(tmp6);
                    auto tmp18 = tmp17 * tmp16;
                    auto tmp19 = tmp18.sin();
                    auto tmp20 = tmp18.cos();
                    tmp19.store(out_ptr1 + static_cast<int64_t>(x0));
                    tmp20.store(out_ptr2 + static_cast<int64_t>(x0));
```

So the question is how we should manage the numeric precision matter of torch.compile. Currently we have all or nothing or fragmented solutions.
1. Global flag like torch._inductor.config.emulate_precision_casts=True can be used (for testing only?) to preserve the cast globally.
2. If we can identify and root cause the area of the code which cause numeric problem, we can use torch._dynamo.graph_break or torch._dynamo disable compile the region of problem.
3. https://github.com/pytorch/pytorch/issues/189334 is proposing a parameter to torch.compile function for specifying precision behanvior.
4. https://github.com/pytorch/pytorch/pull/187971 is adding a global flag torch._inductor.config.numerics = "strict" for makes torch.sum reductions bit-for-bit identical to eager

Usually the code region which is prone to larger numeric drift has a pattern (data cast followed by non-linear operations) and it is constrained to some specific layers or regions. These all or nothing global flag are not ideal to be used in production.  While the graph break or disable torch.compile solutions looks like a hack or a walkaround.

The ideally approach to the torch.compile numeric precision behavior is to design a scoped precision related context which makes the compiling process follow a specified precision behaviors within the context, and back to default (performant) mode when exit from the context.  The scoped precision context can be torch.compile independent which means it is part of the model script no matter whether the model is torch.compiled or not . And it takes effect when it is under the compiled path.

For example:

```
with precision_context(precision_casts=True, numerics="strict", ...):
        temb = time_text_embed

```

Please triage whether this request is reasonable and possible to implement.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo
